firedrake/slate/slac.py

```diff
@@ -0,0 +1,512 @@
+        arglist.append(ast.Decl(ctype, coeffmap[c]))
+
+    if need_cell_facets:
+        arglist.append(ast.Decl("char *", cellfacetsym))
+
+    kernel = ast.FunDecl("void", "slac_compile_slate", arglist,
+                         ast.Block(statements),
+                         pred=["static", "inline"])
+
+    # Finally we transform the kernel into a set of C++ template functions.
+    # This produces the final kernel AST
+    klist = []
+    transformkernel = Transformer()
+    oriented = False
+    for v in kernel_exprs.values():
+        for ks in v:
+            oriented = oriented or ks.kinfo.oriented
+            # TODO: Think about this. Is this true for SLATE?
+            assert ks.kinfo.subdomain_id == "otherwise"
+            kast = transformkernel.visit(ks.kinfo.kernel._ast)
+            klist.append(kast)
+
+    klist.append(kernel)
+    kernelast = ast.Node(klist)
+
+    # Need to get the correct PETSc directory
+    inc.append(petsc.get_petsc_dir() + '/include/eigen3/')
+
+    # Produce the op2 kernel object for assembly
+    op2kernel = op2.Kernel(kernelast,
+                           "slac_compile_slate",
+                           cpp=True,
+                           include_dirs=inc,
+                           headers=['#include <Eigen/Dense>', '#define restrict __restrict'])
+
+    # TODO: What happens for multiple ufl domains?
+    assert len(slate_expr.ufl_domains()) == 1
+    kinfo = KernelInfo(kernel=op2kernel,
+                       integral_type="cell",
+                       oriented=oriented,
+                       subdomain_id="otherwise",
+                       domain_number=0,
+                       coefficient_map=range(len(coeffs)),
+                       needs_cell_facets=need_cell_facets)
+    idx = tuple([0]*len(slate_expr.arguments()))
+
+    return (SplitKernel(idx, kinfo), )
+
+
+def get_c_str(expr, temps, prec=None):
+    """Translates a SLATE expression into its equivalent
+    representation in C. This is done by using COFFEE's
+    code generation functionality and we construct an
+    appropriate C/C++ representation of the `slate.Tensor`
+    expression.
+
+    :arg expr: a :class:`slate.TensorBase` expression.
+
+    :arg temps: a `dict` of temporaries which map a given
+                `slate.Tensor` object to its corresponding
+                representation as a `coffee.Symbol` object.
+
+    :arg prec: an argument dictating the order of precedence
+               in the linear algebra operations. This ensures
+               that parentheticals are placed appropriately
+               and the order in which linear algebra operations
+               are performed are correct.
+
+    Returns
+        This function returns a `string` which represents the
+        C/C++ code representation of the `slate.Tensor` expr.
+    """
+    if isinstance(expr, Tensor):
+        return temps[expr].gencode()
+
+    elif isinstance(expr, UnaryOp):
+        if expr.op == "Transpose":
```

Comment:
I mean use `instance(expr, Transpose)` instead of `expr.op == "Transpose"`. You can also go `singledispatch`, that way you don't have to topologically sort if...elif...else... branches.